apps/e2e/ag-grid-storybook-e2e/cypress.config.ts

```diff
@@ -1,21 +1,85 @@
+import * as utils from '@percy/sdk-utils';
 
 import { defineConfig } from 'cypress';
+import * as fs from 'fs';
 
-const cypressJsonConfig = {
-  fileServerFolder: '.',
-  fixturesFolder: './src/fixtures',
-  video: false,
-  videosFolder: '../../../dist/cypress/apps/e2e/ag-grid-storybook-e2e/videos',
-  screenshotsFolder:
-    '../../../dist/cypress/apps/e2e/ag-grid-storybook-e2e/screenshots',
-  chromeWebSecurity: false,
-  specPattern: 'src/e2e/**/*.cy.{js,jsx,ts,tsx}',
-  supportFile: 'src/support/e2e.ts',
-};
 export default defineConfig({
   e2e: {
     ...nxE2EPreset(__dirname),
-    ...cypressJsonConfig,
+    video: false,
+    setupNodeEvents: (on, config) => {
+      // Use a large window size so screenshots fit within the viewport.
+      on('before:browser:launch', (browser, launchOptions) => {
+        const width = 1920;
+        const height = 1600;
+
+        if (browser.name === 'chrome' && browser.isHeadless) {
+          launchOptions.args.push(`--window-size=${width},${height}`);
+          launchOptions.args.push('--force-device-scale-factor=2');
+          launchOptions.args.push('--no-sandbox');
+          launchOptions.args.push('--disable-setuid-sandbox');
+        }
+
+        if (browser.name === 'electron' && browser.isHeadless) {
+          launchOptions.preferences.width = width;
+          launchOptions.preferences.height = height;
+          launchOptions.preferences.deviceScaleFactor = 2;
+        }
+
+        if (browser.name === 'firefox' && browser.isHeadless) {
+          launchOptions.args.push(`--window-size=${width},${height}`);
+        }
+
+        return launchOptions;
+      });
+
+      on(
+        'after:screenshot',
+        async (
+          details: Cypress.ScreenshotDetails
+        ): Promise<Cypress.AfterScreenshotReturnObject> => {
+          const url = (
+            details.blackout.find((item) => item.startsWith('url:')) || ''
+          ).substring(4);
+          if (process.env.PERCY_SERVER_ADDRESS && url) {
+            const imageDataBase64 = fs
+              .readFileSync(details.path)
+              .toString('base64');
+            const { width, height } = details.dimensions;
+            const domSnapshot = `<!doctype html>
+            <html lang="en">
+              <head>
+                <meta charset="utf-8">
+                <title>${details.name}</title>
+                <style>
+                  *, *::before, *::after { margin: 0; padding: 0; font-size: 0; }
+                  html, body { width: 100%; }
+                  img { max-width: 100%; }
+                </style>
+              </head>
+              <body>
+                <img
+                  alt="${details.name}"
+                  src="data:image/png;base64,${imageDataBase64}"
+                  width="${width / 2}"
+                  style="aspect-ratio: ${width}/${height};"
+                />
+              </body>
+            </html>`;
```

Comment:
This HTML wrapper for the image works like `npx percy upload`, but having our own copy allows control over the snapshot names. Inlining the image source is easier than managing multiple resources.

Reply:
This comment would be excellent to have in the source code, too.